router/src/block_allocator.rs

```diff
@@ -0,0 +1,128 @@
+    block_allocator: mpsc::UnboundedSender<BlockAllocatorCommand>,
+}
+
+impl BlockAllocator {
+    pub(crate) fn new(
+        max_batch_total_tokens: u32,
+        block_size: u32,
+        window_size: Option<u32>,
+    ) -> Self {
+        // Create channel
+        let (sender, receiver) = mpsc::unbounded_channel();
+
+        // Launch background queue task
+        tokio::spawn(block_allocator_task(
+            max_batch_total_tokens / block_size,
+            block_size,
+            window_size,
+            receiver,
+        ));
+
+        Self {
+            block_allocator: sender,
+        }
+    }
+
+    pub(crate) async fn allocate(&self, tokens: u32) -> Option<BlockAllocation> {
+        let (response_sender, response_receiver) = oneshot::channel();
+        self.block_allocator
+            .send(BlockAllocatorCommand::Allocate {
+                tokens,
+                response_sender,
+            })
+            .unwrap();
+
+        response_receiver
+            .await
+            .unwrap()
+            .map(|(blocks, slots)| BlockAllocation {
+                blocks,
+                slots,
+                block_allocator: self.clone(),
+            })
+    }
+
+    pub(crate) fn free(&self, blocks: Vec<u32>) {
+        self.block_allocator
+            .send(BlockAllocatorCommand::Free { blocks })
+            .unwrap();
+    }
+}
+
+async fn block_allocator_task(
+    blocks: u32,
+    block_size: u32,
+    window_size: Option<u32>,
+    mut receiver: mpsc::UnboundedReceiver<BlockAllocatorCommand>,
+) {
+    // Block 0 is reserved for health checks
+    let mut free_blocks: Vec<u32> = (1..blocks).collect();
+    while let Some(cmd) = receiver.recv().await {
+        match cmd {
+            BlockAllocatorCommand::Free { blocks } => free_blocks.extend(blocks),
+            BlockAllocatorCommand::Allocate {
+                tokens,
+                response_sender,
+            } => {
+                // Apply window size
+                let (tokens, repeats) = match window_size {
+                    None => (tokens, 1),
+                    Some(window_size) => {
+                        let repeats = (tokens + window_size - 1) / window_size;
+                        let tokens = min(tokens, window_size);
+                        (tokens, repeats as usize)
+                    }
+                };
+
+                // Pad to a multiple of block size
+                let required_blocks = (tokens + block_size - 1) / block_size;
+                if required_blocks > free_blocks.len() as u32 {
+                    response_sender.send(None).unwrap();
+                } else {
+                    let mut blocks = Vec::with_capacity(required_blocks as usize);
+                    let mut slots = Vec::with_capacity((required_blocks * block_size) as usize);
+
+                    for _ in 0..required_blocks {
+                        let block_id = free_blocks.pop().unwrap();
```

Comment:
```suggestion
                        let block_id = free_blocks.pop().expect("Having free slots');
```
Feels like a pretty bad panic we'd really like to know where it comes from, no ?

Reply:
it's unreachable since we check the length of the vec beforehand.

Reply:
Fair enough